backend/tests/unit/protocols/server.txt

```diff
@@ -0,0 +1,50 @@
+Scenario: The server's Frontendconnection event correctly reports whether it has data
+Given: A Frontendconnection event object is constructed with default parameters
+When: The object is not modified
+Then: has_data function returns false
+And_When: A new value is set for the time field of the object
+Then: has_data function returns true
+
+Scenario: The server receives various input events properly and reports whether event has data
+Given: A ReceiveEvent object is constructed with default parameters
+When: A new value is set for the time field of the object
+Then: has_data function returns true
+
+Scenario: The server receives output event and reports correctly whether event has data
+Given: A ReceiveOutputEvent object is constructed with default parameters
+When: The object is not modified
+Then: has_data function returns false
+And_When: The server receives output event with data
+Then: has_data function returns true
+
+Scenario: The server sends output event properly and reports whether event has data
+Given: A SendOutputEvent object is constructed with default parameters
+When: The data is sent via serial
+Then: has_data function returns true
+
+Scenario: The make_serial_receive function creates a ReceiveEvent object correctly
+Given: The make_serial_receive function
+When: The data received from serial is given to the function
```

Comment:
These `make_*` functions should make zero assumptions about where the data is coming from, so we should remove those assumptions from the Whens. It would be clearer and more specific to describe exactly what assumptions are made about the input parameters (e.g does the parameter need to be non-`None`? Does it need to be a bytestring?)